libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/roadobject/RoadObjectEx.kt

```diff
@@ -2,68 +2,66 @@
 
 package com.mapbox.navigation.core.trip.roadobject
 
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.navigation.base.internal.route.nativeRoute
 import com.mapbox.navigation.base.route.NavigationRoute
 import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.base.trip.model.roadobject.RoadObjectType
 import com.mapbox.navigation.base.trip.model.roadobject.UpcomingRoadObject
-import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFactory
-import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromIntersection
-import com.mapbox.navigation.core.trip.roadobject.reststop.RestStopFromRoadObject
+import com.mapbox.navigation.core.navigator.toUpcomingRoadObjects
 import com.mapbox.navigation.utils.internal.ifNonNull
+import com.mapbox.navigator.RouteInterface
 
 /**
- * Use the API to retrieve a list of all upcoming rest stops using [UpcomingRoadObject].
- * This is recommended to be used with
- * - [RouteProgress] during Active Navigation.
+ * Use the API to retrieve the first upcoming [UpcomingRoadObject].
+ * - given that route has no [UpcomingRoadObject] of type [clazz], returns null
+ * - given that route has 1 [UpcomingRoadObject] of type [clazz], returns it considering
+ *   [UpcomingRoadObject.distanceToStart] is positive
+ * - given that route has 1 [UpcomingRoadObject] of type [clazz], returns null considering
+ *   [UpcomingRoadObject.distanceToStart] is either 0, negative or null
+ * - given that route has multiple upcoming road objects of type [clazz], returns the first
```

Comment:
But it doesn't work like this. We look for the first item with the correct type and then either return it or null, depending on whether distanceToStart is positive or not.